packages/internal/src/debug.js

```diff
@@ -1,41 +1,108 @@
 // @jessie-check
 
+/**
+ * @import {OptPrefixPath} from './types-internal.js';
+ */
+
+// TODO This is top level static mutable state. Why did it pass at-jessie-check?
 let debugInstance = 1;
 
+const logDisabled = (..._args) => {};
+logDisabled.sub = _ => logDisabled;
+harden(logDisabled);
+
 /**
- * @param {string} name
+ * @typedef {typeof logDisabled} TraceLogger
+ */
+
+/**
+ * @param {OptPrefixPath} optPrefixPath
+ * @param {number} rootInstanceNum
  * @param {boolean | 'verbose'} enable
+ * @returns {TraceLogger}
  */
-export const makeTracer = (name, enable = true) => {
-  debugInstance += 1;
+const makeSubTracer = (optPrefixPath, rootInstanceNum, enable) => {
+  // DO NOT ASSIGN to the parameter variables above, since they are captured
+  // by closures below.
+  const makeKey = () => {
+    // As much as possible, postpone anything expensive to here, since this is
+    // only called when a tracer or sub-tracer function is called. Not when it
+    // is made.
+    let prefix;
+    // instead, make a local assignable copy
+    let optPath = optPrefixPath;
+    while (optPath !== undefined) {
+      const [step, optRestSteps] = optPath;
+      if (prefix === undefined) {
+        prefix = step;
+      } else {
+        prefix = `${step}.${prefix}`;
+      }
+      optPath = optRestSteps;
+    }
+    return `----- ${prefix}.${rootInstanceNum} `;
+  };
+  const sub = prefixStep =>
+    makeSubTracer([prefixStep, optPrefixPath], rootInstanceNum, enable);
+
+  // Because `debugCount` is pre-incremented before output, starting it at
+  // `1` means the output counts start at `2`.
   let debugCount = 1;
-  const key = `----- ${name}.${debugInstance} `;
   // the cases below define a named variable to provide better debug info
   switch (enable) {
     case false: {
-      const logDisabled = (..._args) => {};
       return logDisabled;
     }
     case 'verbose': {
       const infoTick = (optLog, ...args) => {
-        if (optLog.log) {
+        const key = makeKey();
+        if (typeof optLog.log === 'function') {
```

Comment:
The implementation of `makeTracer` properly relies on top-level mutable state, only in order to produce console log output that only privileged parties can see. ***HOWEVER***, by passing in one's own alleged logger object (object with a `.log` method), one can read the `debugInstance` top level mutable state, enabling subsystems that should be isolated from each other to communicate without using any special privilege, such as reading console output.

Reply:
Resolving, since we decided against making this change, instead documenting the security hazard at #11845 .

@dckc after reading your example code again, at #11857 I did change to your `Object.assign` pattern after all.